modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServicesDeploymentTask.java

```diff
@@ -0,0 +1,856 @@
+
+                    final Map<IgniteUuid, ServiceInfo> deployedServices = srvcProc.deployedServices();
+
+                    if (evtType == EVT_NODE_LEFT || evtType == EVT_NODE_FAILED) {
+                        deployedServices.forEach((srvcId, desc) -> {
+                            if (desc.topologySnapshot().containsKey(eventNode.id()) ||
+                                (desc.cacheName() != null && !eventNode.isClient())) // If affinity service
+                                toDeploy.put(srvcId, desc);
+                        });
+                    }
+                    else {
+                        toDeploy.putAll(deployedServices);
+
+                        toDeploy.putAll(srvcProc.servicesReceivedFromJoin(eventNode.id()));
+                    }
+                }
+
+                if (toDeploy.isEmpty()) {
+                    completeSuccess();
+
+                    if (log.isDebugEnabled())
+                        log.debug("No services deployment deployment action required.");
+
+                    return;
+                }
+
+                depActions = new ServicesDeploymentActions();
+
+                depActions.servicesToDeploy(toDeploy);
+            }
+
+            ClusterNode crd = srvcProc.coordinator();
+
+            if (crd == null) {
+                onAllServersLeft();
+
+                return;
+            }
+
+            crdId = crd.id();
+
+            if (crd.isLocal())
+                initCoordinator(evtTopVer);
+
+            processDeploymentActions(depActions);
+        }
+        catch (Exception e) {
+            log.error("Error occurred while initializing deployment task, err=" + e.getMessage(), e);
+
+            completeError(e);
+
+            throw new IgniteCheckedException(e);
+        }
+        finally {
+            if (!initTaskFut.isDone())
+                initTaskFut.onDone();
+
+            if (log.isDebugEnabled()) {
+                log.debug("Finished services deployment future init: [depId=" + deploymentId() +
+                    ", locId=" + ctx.localNodeId() + ']');
+            }
+        }
+    }
+
+    /**
+     * @param depActions Services deployment actions.
+     */
+    @SuppressWarnings("ErrorNotRethrown")
+    private void processDeploymentActions(@NotNull ServicesDeploymentActions depActions) {
+        srvcProc.updateDeployedServices(depActions);
+
+        depActions.servicesToUndeploy().forEach((srvcId, desc) -> {
+            srvcProc.deployment().deployerBlockingSectionBegin();
+
+            try {
+                srvcProc.undeploy(srvcId);
+            }
+            finally {
+                srvcProc.deployment().deployerBlockingSectionEnd();
+            }
+        });
+
+        depActions.servicesToDeploy().forEach((srvcId, desc) -> {
+            try {
+                ServiceConfiguration cfg = desc.configuration();
+
+                Map<UUID, Integer> top = reassign(srvcId, cfg, evtTopVer, filterDeadNodes(desc.topologySnapshot()));
```

Comment:
This code is executed in `ServicesDeploymentWorker` thread, so discovery events may occur concurrently with it.
So, it's possible for different nodes to get different `oldTop` and calculate different service assignments.

It would be great to cover such cases with tests. Try reproducing this bug in a test.

Reply:
Fixed.